File: Lab3/src/KI303/Krupskyi/Lab3/Phone.java
package KI303.Krupskyi.Lab3;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;

/**
 * Абстрактний клас {@code Phone} моделює базову предметну область "Телефон".
 * Містить складові частини — екран, процесор і батарею, 
 * а також базові методи для імітації роботи пристрою.
 * Усі події логуються у файл {@code phone_log.txt}.
 *
 * @author Крупський Ілля
 * @version 1.0
 * @since 2025-10-23
 */
public abstract class Phone {

    /** Вкладений клас, що описує екран телефону */
    class Display {
        double size;
        String resolution;

        Display(double size, String resolution) {
            this.size = size;
            this.resolution = resolution;
        }

        @Override
        public String toString() {
            return "Display: " + size + "\" " + resolution;
        }
    }

    /** Вкладений клас, що описує процесор телефону */
    class Processor {
        String model;
        int cores;

        Processor(String model, int cores) {
            this.model = model;
            this.cores = cores;
        }

        @Override
        public String toString() {
            return "Processor: " + model + " (" + cores + " cores)";
        }
    }

    /** Вкладений клас, що описує батарею телефону */
    class Battery {
        int capacity;
        int charge;

        Battery(int capacity, int charge) {
            this.capacity = capacity;
            this.charge = charge;
        }

        void chargeBattery(int amount) {
            charge = Math.min(100, charge + amount);
        }

        void discharge(int amount) {
            charge = Math.max(0, charge - amount);
        }

        @Override
        public String toString() {
            return "Battery: " + capacity + " mAh, charge = " + charge + "%";
        }
    }

    // --- Основні поля ---
    protected String brand;
    protected Display display;
    protected Processor processor;
    protected Battery battery;
    protected boolean poweredOn;
    protected PrintWriter log;

    // --- Конструктор ---
    /**
     * Конструктор із параметрами.
     * @throws IOException якщо не вдається створити лог-файл
     */
    public Phone(String brand, double screenSize, String resolution,
                 String cpu, int cores, int batteryCap, int charge) throws IOException {
        this.brand = brand;
        this.display = new Display(screenSize, resolution);
        this.processor = new Processor(cpu, cores);
        this.battery = new Battery(batteryCap, charge);
        this.poweredOn = false;
        this.log = new PrintWriter(new FileWriter("phone_log.txt", true));
        log.println("Created phone: " + brand);
        log.flush();
    }

    /** Увімкнути телефон */
    public void powerOn() {
        if (!poweredOn) {
            poweredOn = true;
            log.println(brand + " powered ON.");
            log.flush();
        }
    }

    /** Вимкнути телефон */
    public void powerOff() {
        if (poweredOn) {
            poweredOn = false;
            log.println(brand + " powered OFF.");
            log.flush();
        }
    }

    /** Зарядити телефон */
    public void chargePhone() {
        battery.chargeBattery(10);
        log.println(brand + " charging... ");
        log.flush();
    }

    /** Показати рівень заряду */
    public int getChargeLevel() {
        return battery.charge;
    }

    /** Отримати модель процесора */
    public String getProcessorModel() {
        return processor.model;
    }

    /** Перевірити, чи увімкнений телефон */
    public boolean isPoweredOn() {
        return poweredOn;
    }

    /** Закрити лог-файл */
    public void closeLog() {
        if (log != null) {
            log.println("Closing log for " + brand);
            log.close();
        }
    }

    /** Абстрактний метод, що реалізується у підкласах */
    public abstract void makeCall(String number);

    @Override
    public String toString() {
        return "Phone: " + brand + "\n" +
                display + "\n" +
                processor + "\n" +
                battery + "\n" +
                "State: " + (poweredOn ? "On" : "Off");
    }
}

File: Lab9/KI303_Krupskyi_Lab9/phone.py
import os

# Базовий клас "Телефон", що описує загальні характеристики
class Phone:
    # Конструктор класу. Ініціалізує властивості телефону
    def __init__(self, brand, screen_size, battery_capacity):
        self.brand = brand                  # Марка телефону
        self.screen_size = screen_size      # Розмір екрану
        self.battery_capacity = battery_capacity # Ємність батареї
        self.charge = 100                   # Початковий рівень заряду (у відсотках)
        
        # Виводимо повідомлення про створення об'єкту
        print(f"Phone created: {self.brand}, Screen: {self.screen_size}, Battery: {self.battery_capacity}")

    # Метод для здійснення дзвінка
    def make_call(self, number):
        # Перевіряємо, чи достатньо заряду для дзвінка
        if self.charge > 5:
            print(f"Calling {number} from {self.brand}...")
            self.charge -= 5  # Зменшуємо заряд на 5%
        else:
            print(f"Not enough battery to call from {self.brand}.")

    # Метод для зарядки телефону
    def charge_battery(self, amount):
        self.charge += amount
        # Обмеження заряду максимумом у 100%
        if self.charge > 100:
            self.charge = 100
        print(f"Battery charged. Current level: {self.charge}%")

    # Метод для отримання поточного рівня заряду (гетер)
    def get_charge_level(self):
        return self.charge